lib/node-loader-babel.js

```diff
@@ -19,15 +19,23 @@ function isBabelConfigFile(filename) {
 
 export async function load(url, context, defaultLoad) {
   if (useLoader(url)) {
-    const { source } = await defaultLoad(url, context, defaultLoad);
+    const { source, format } = await defaultLoad(url, context, defaultLoad);
+
+    // Skip transpilation of CommonJS modules.
+    // These modules are already preprocessed by Node.js,
+    // so we cannot parse the non-standard syntaxes like JSX and TypeScript.
+    // Their transpilation is better handled separately by @babel/register or @babel/node.
+    if (format !== "module") {
+      return { source, format };
+    }
```

Comment:
I don't think that skipping compilation of commonjs modules is a good universal assumption, since babel is completely capable of compiling commonjs modules. I agree skipping json, wasm, and builtin formats makes sense. As a separate feature, I think that we could let users of node-loader-babel customize which files are compiled by providing their own useLoader function. But that wouldn't be part of this PR.

As far as I know, Babel doesn't change the format of modules by default (unless you add a plugin that does it). Babel preset env doesn't seem to do so by default (see [this repl](https://babeljs.io/repl/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=KYDwDg9gTgLgzgOgBbADaogAgLyYORICWeA3EA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.16.6&externalPlugins=&assumptions=%7B%7D)). So I would think it's safe to let babel just compile the file like normal?

```suggestion
    if (format !== "module" && format !== "commonjs") {
      return { source, format };
    }
```

Reply:
> I don't think that skipping compilation of commonjs modules is a good universal assumption, since babel is completely capable of compiling commonjs modules. I agree skipping json, wasm, and builtin formats makes sense. As a separate feature, I think that we could let users of node-loader-babel customize which files are compiled by providing their own useLoader function. But that wouldn't be part of this PR.

Looking forward what you say makes sense. It may be that one day node loader handles the `source` key for `{format: 'commonjs'}`. But insisting on `commonjs` being transpiled now makes little sense. As far as I understand, with the code of this PR, a commonjs input will have `const {format, source} = await defaultLoad(...);` such that `format = 'commonjs'` and `source = null`. Babel gracefully handles `null` by producing `null` with little overhead. If one day `source` is non-null and has meaning for `format = 'commonjs'` suddenly this loader will start transpiling CJS input without warning. Not sure if this is a real stability threat but I kind of feel like it is a bad idea to leave this open for this code to be general just in the case if.

I agree this behavior could be configurable by the user of `node-loader-babel`. I just think the default should be to only process `'module'` input.

Note that without this PR, source is `null` and output format is `module` which is the reason of the bug described in #7.

Reply:
> As far as I know, Babel doesn't change the format of modules by default (unless you add a plugin that does it). Babel preset env doesn't seem to do so by default (see [this repl](https://babeljs.io/repl/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=KYDwDg9gTgLgzgOgBbADaogAgLyYORICWeA3EA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.16.6&externalPlugins=&assumptions=%7B%7D)). So I would think it's safe to let babel just compile the file like normal?

In this REPL, enabling `@babel/preset-env` does transform ESM input into CJS output. How do you configure it so that it does not change format?

Reply:
> In this REPL, enabling @babel/preset-env does transform ESM input into CJS output. How do you configure it so that it does not change format?

Set babel preset env's `modules` option to `false`. The `"auto"` value also should work if we have called babel with the proper options

https://babeljs.io/docs/en/babel-preset-env#modules

> I just think the default should be to only process 'module' input.

I can see how that would help solve the immediate problem, but don't think it's actually the proper behavior for node-loader-babel as a whole. If we can get babel to output the correct module format, there's no reason to avoid compilation of CJS files.

> with the code of this PR, a commonjs input will have const {format, source} = await defaultLoad(...); such that format = 'commonjs' and source = null

Why would `source` be `null`? Does nodejs' default loader not provide commonjs source as a string? If so, then perhaps your suggestion of disabling would make more sense. But I thought that nodejs would give us the source as a string.

Reply:
OK. I cannot find how to do this in the REPL, maybe it is not possible.